formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/FormstackPerformSarHandler.scala

```diff
@@ -0,0 +1,87 @@
+
+import java.time.{Instant, LocalDate, LocalDateTime}
+
+import com.gu.identity.formstackbatonrequests.BatonModels.{SarRequest, SarResponse}
+import com.gu.identity.formstackbatonrequests.aws.{DynamoClient, SubmissionTableUpdateDate}
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.concurrent.ExecutionContext
+
+case class SubmissionIdEmail(email: String, submissionId: String, receivedByLambdaTimestamp: Long, accountNumber: Int)
+
+case class FormstackPerformSarHandler(
+  dynamoClient: DynamoClient,
+  formstackClient: FormstackSar,
+  config: PerformSarLambdaConfig)
+  extends LazyLogging with FormstackHandler[SarRequest, SarResponse] {
+
+  implicit val ec: ExecutionContext = ExecutionContext.global
+
+  def submissionsWithEmailAndAccount(submissions: List[FormstackSubmission], accountNumber: Int): List[SubmissionIdEmail] = {
+    val emailReg = """(?i)\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}\b""".r
+    submissions.foldLeft(List.empty[SubmissionIdEmail]) { (acc, submission) =>
+      val submissionValues = submission.data.map(field => field._2.value).toList
+      val emailList = submissionValues.collect { case jsonValue => emailReg.findAllIn(jsonValue.toString).toList }.flatten
+      val receivedByLambdaTimestamp = Instant.now.getEpochSecond
+      val submissionsIdEmails = emailList.map(email => SubmissionIdEmail(email.toLowerCase, submission.id, receivedByLambdaTimestamp, accountNumber))
+      submissionsIdEmails ::: acc
+    }
+  }
+
+  private def skipDecryptionError(err: Throwable): Either[Throwable, Unit] = {
+    err match {
+      case _: FormstackDecryptionError => Right(())
+      case err => Left(err)
+    }
+  }
+
+  private def handleSubs(form: FormstackForm, lastUpdate: SubmissionTableUpdateDate, submissionPage: Int = 1, token: FormstackAccountToken): Either[Throwable, Unit] = {
+    formstackClient.formSubmissionsForGivenPage(submissionPage, form.id, lastUpdate, config.encryptionPassword, token) match {
+      case Left(err) => skipDecryptionError(err)
+      case Right(response) =>
+        logger.info(s"Received page $submissionPage of submissions out of ${response.pages} pages for form ${form.id}.")
+        val submissionsIdsWithEmails = submissionsWithEmailAndAccount(response.submissions, token.account)
+        logger.info(s"Writing ${submissionsIdsWithEmails.length} submission id(s) and emails to Dynamo")
+        val writeResults = dynamoClient.writeSubmissions(submissionsIdsWithEmails, config.bcryptSalt)
+        val errors = writeResults.collect { case Left(err) => err }
+        if (errors.nonEmpty) {
+          Left(new Exception(errors.toString()))
+        } else if (submissionPage < response.pages) {
+          handleSubs(form, lastUpdate, submissionPage + 1, token)
+        } else Right(())
+    }
+  }
+
+  def updateSubmissionsTable(formsPage: Int, lastUpdate: SubmissionTableUpdateDate, count: Int, token: FormstackAccountToken): Either[Throwable, Unit] = {
```

Comment:
This method and the one above are both tail recursive. This enables us to process the paginated responses returned by Formstack